turbine/src/retransmit_stage.rs

```diff
@@ -370,7 +405,106 @@ fn retransmit_shred(
     stats
         .retransmit_total
         .fetch_add(retransmit_time.as_us(), Ordering::Relaxed);
-    Some((key.slot(), root_distance, num_nodes))
+    Some(RetransmitShredOutput {
+        shred: key,
+        last_shred_in_slot,
+        root_distance,
+        num_nodes,
+        addrs: match addrs {
+            Cow::Owned(addrs) => Some(addrs.into_boxed_slice()),
+            Cow::Borrowed(_) => None,
+        },
+    })
+}
+
+fn get_retransmit_addrs<'a>(
+    shred: &ShredId,
+    root_bank: &Bank,
+    cache: &HashMap<Slot, (/*leader:*/ Pubkey, Arc<ClusterNodes<RetransmitStage>>)>,
+    addr_cache: &'a AddrCache,
+    socket_addr_space: &SocketAddrSpace,
+    stats: &RetransmitStats,
+) -> Option<(/*root_distance:*/ u8, Cow<'a, [SocketAddr]>)> {
+    if let Some((root_distance, addrs)) = addr_cache.get(shred) {
```

Comment:
At this point can we also remove `shred` from the cache?

Reply:
We might need to keep it due to duplicate shreds.
Also, this is called within a thread-pool, so mutations at least need some locking etc.